src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -163,7 +163,7 @@ int MPIR_Typerep_unpack(const void *inbuf, MPI_Aint insize, void *outbuf, MPI_Ai
     int mpi_errno = MPI_SUCCESS;
     int rc;
 
-    if (outcount == 0) {
+    if (insize == 0) {
```

Comment:
How about `outcount == 0 || insize == 0`?

Reply:
`outcount == 0` is kind of a strange case, IMO. If we have data to unpack, unless there is truncation, count should be > 0. I'm not sure it is worth guarding against.

Reply:
I kind of agree, from integral point of view. But if viewed from `Typerep` alone, then as a component, it should be prepared for all cases so the correctness of the component doesn't depend on the input. I think it has a merit of a defensive measure.

Reply:
Yes, but it adds a branch if we compare both values. Since this is an internal API, I think it is OK to ignore what is erroneous usage and save the branch.

Reply:
> Yes, but it adds a branch if we compare both values. 

Yes. However, if you believe a condition will never arise, then the cost of that branch is zero, right?

Reply:
The compiler does not know that it will never arise. I guess we could mark it as unlikely as a hint. This is the unpack case, so single branch additions probably don't make much of a difference...

Reply:
I am reading this blog: https://xania.org/201602/bpu-part-one, it seems --

It depends on whether the CPU use static prediction or dynamic prediction. The latter is used in Intel CPU since Sandybridge (I just learned this myself 😄 ). With dynamic prediction, the CPU counter itself should be able to eliminate the branch cost if the code has been repeated a few times and each time only take one branch. 

The compiler can take the hints and arrange the code to help the CPU for the cases initially or the CPU counter get reset (due to branch uncertainty), but for branches that is always false or true, I think it really doesn't matter.

I don't think we want to add the `unlikely` cruft everywhere until we are convinced it is necessary.

Reply:
> I don't think we want to add the `unlikely` cruft everywhere until we are convinced it is necessary.

I would say the same about the extra branch not being necessary.

Reply:
OK. I'll leave the decision to you.